jcl/src/openj9.sharedclasses/share/classes/com/ibm/oti/shared/SharedClassURLClasspathHelperImpl.java

```diff
@@ -166,53 +177,70 @@ public synchronized boolean storeSharedClass(String partition, Class<?> clazz, i
-			printVerboseError(Msg.getString("K05a8")); //$NON-NLS-1$
-			return false;
-		}
-		if (invalidURLExists) {
-			/* Any URL which has its protocol other than 'jar:' or 'file:' is not supported by
-			 * shared class cache and is considered invalid.
-			 * invalidURLExists = true indicates classpath contains an invalid URL,
-			 * As such there is no point in calling native method storeSharedClassImpl2() 
-			 * since it is bound to fail when creating classpath entries.
-			 */			
-			/*[MSG "K05a9", "Classpath contains an invalid URL. Returning false."]*/
-			printVerboseInfo(Msg.getString("K05a9")); //$NON-NLS-1$
+		ClassLoader actualLoader = getClassLoader();
+		if (!validateClassLoader(actualLoader, clazz)) {
+			/* Attempt to call storeSharedClass with class defined by a different classloader */
 			return false;
 		}
-		if (!validated[foundAtIndex]) {
-			/* Because we only check each element once, we can afford to also check whether the URL exists */
-			if (!validateURL(urls[foundAtIndex], true)) {
+		boolean storeRet = false;
+		boolean incConfirmedCount = false;
+		boolean urlValidated = false;
+		urlcpReadWriteLock.readLock().lock();
+		try {
+			if (urlCount==0) {
+				/*[MSG "K05a6", "Classpath has zero elements. Cannot call storeSharedClass without classpath. Returning false."]*/
+				printVerboseError(Msg.getString("K05a6")); //$NON-NLS-1$
 				return false;
 			}
-			validated[foundAtIndex]=true;
+			if (foundAtIndex>=urlCount) {
+				/*[MSG "K05a8", "foundAtIndex cannot be >= urls in classpath for storeSharedClass. Returning false."]*/
+				printVerboseError(Msg.getString("K05a8")); //$NON-NLS-1$
+				return false;
+			}
+			if (invalidURLExists) {
+				/* Any URL which has its protocol other than 'jar:' or 'file:' is not supported by
+				 * shared class cache and is considered invalid.
+				 * invalidURLExists = true indicates classpath contains an invalid URL,
+				 * As such there is no point in calling native method storeSharedClassImpl2() 
+				 * since it is bound to fail when creating classpath entries.
+				 */			
+				/*[MSG "K05a9", "Classpath contains an invalid URL. Returning false."]*/
+				printVerboseInfo(Msg.getString("K05a9")); //$NON-NLS-1$
+				return false;
+			}
+			if (!validated[foundAtIndex]) {
+				/* Because we only check each element once, we can afford to also check whether the URL exists */
+				if (!validateURL(urls[foundAtIndex], true)) {
+					return false;
+				} 
+				urlValidated = true;
+			}
+			if (confirmedCount <= foundAtIndex) {
+				incConfirmedCount = true;
+			}
+			storeRet = storeSharedClassImpl2(this.id, partition, actualLoader, this.urls, this.urlCount, foundAtIndex, clazz, nativeFlags);
+		} finally {
+			urlcpReadWriteLock.readLock().unlock();
 		}
-		ClassLoader actualLoader = getClassLoader();
-		if (!validateClassLoader(actualLoader, clazz)) {
-			/* Attempt to call storeSharedClass with class defined by a different classloader */
-			return false;
+		if (urlValidated) {
```

Comment:
Might as well combine updating validated and confirmedCount to use a single writeLock().lock().

Reply:
I added `setValidatedAtIndex()`, but I am wondering if we really need that. I guess the previous code that sets `validated[foundAtIndex]=true;` inside readLock should be fine, we just may have multiple threads setting it to true, which won't have any negative impact.

Reply:
I'm ok with that.

Reply:
`setValidatedAtIndex()` removed.